Modelica/Electrical/Analog/Lines/Functions/lineZmatrix.mo

```diff
@@ -0,0 +1,116 @@
+function lineZmatrix
+  "Compute matrix of longitudinal impedances per metre of a multi-conductor line"
+  extends Modelica.Icons.Function;
+  import Modelica.ComplexMath;
+  import Modelica.Utilities.Streams;
+  import Modelica.Constants.pi;
+  import Modelica.Constants.mu_0;
+  import Modelica.ComplexMath.j;
+
+  input Integer n "Number of conductors in line";
+  input Modelica.Units.SI.Length x[n] "Horizontal abscissas of conductors";
+  input Modelica.Units.SI.Length y[n] "Vertical abscissas of conductors";
+  input Modelica.Units.SI.Length r[n] "Conductors radii";
+  input Modelica.Units.SI.ResistancePerUnitLength R1[n] "Conductors lineic resistance";
+  input Real k_s[n]=fill(0.7,n) "Ratio of equivalent shell radius to actual radius";
+  //  in case of cylindric conductor this is equal to exp(-mu_r/4)=exp(-0.25)
+  input Modelica.Units.SI.Resistivity rho=100 "Ground resistivity";
+  input Modelica.Units.SI.Frequency f=50 "Frequency";
+  output Modelica.Units.SI.ResistancePerUnitLength Rcomp[div(n * (n + 1), 2)] "Compact resistance matrix";
+  output Modelica.Units.SI.ReactancePerUnitLength Xcomp[div(n * (n + 1), 2)] "Compact reactance matrix";
+  output Modelica.Units.SI.InductancePerUnitLength Lcomp[div(n * (n + 1), 2)] "Compact inductance)";
+protected
+  Modelica.Units.SI.Distance D "Generic larger distance";
+  Modelica.Units.SI.Distance d "Generic smaller distance";
+  Real Theta "Theta angle for Carson's formulas";
+  Real L "generic inductance";
+  Real P "P coefficient from Carson's original paper";
+  Real Q "Q coefficient from Carson's original paper";
+
+  Complex Z[n, n] "Computed matrix (ohm/m)";
+
+  final parameter Real w=2*pi*f;
+  parameter Real a0=sqrt(w*mu_0/rho) "multiplies D in a-parameter from EMTPs Theory Book";
+  // a0 is written, equivalently, as 4*pi*sqrt(5)*1e-4*sqrt(f/rho) in EMTP Theory book
+  //(reference available in Electrical.Analog.References)
+  //
+  //  a0*D=a; a is the same as "r" in Carson's paper
+  // where D=2h for self impedance, D=Dik for mutual impedance
+  Real a;
+  Integer k;
+algorithm
+//self impedances with ground return:
+  for i in 1:n loop
+    a:=a0*2*y[i];
+    assert(a<0.25,"parameter a="+String(a)+
+                  " >0.25 will result in inadequate precision of Carson's formulas",
+                  AssertionLevel.warning);
+    P := pi/8 - sqrt(2)/6*a + a^2/16*(0.672784 + log(2/a));
+    Q := (-0.03860784) + 0.5*log(2/a) + sqrt(2)/6*a-pi/64*a*a;
+    L := mu_0/(2*pi)*log(2*y[i]/(k_s[i]*r[i])) "L when soil is pure conductive";
+    Z[i, i] := R1[i] + 4*w*P/1.e7 + j*w*(L + 4*Q/1.e7); //Carson (25) and (30)
+  end for;
+//mutual impedances with ground return:
+  for i in 1:n loop
+    for jj in 1:i - 1 loop
+      Theta := atan(abs((x[i] - x[jj])/(y[i] + y[jj])));
+      d := sqrt((x[i] - x[jj])^2 + (y[i] - y[jj])^2); //distanza fra conduttore i e jj
+      D := sqrt((x[i] - x[jj])^2 + (y[i] + y[jj])^2);  //distanza fra conduttore i e immagine di jj
```

Comment:
Nice to read Italian here but I fear we shall switch to English. ;-)

Reply:
fixed